### DISCUSSION TITLE
What could be done with "plugins"?

### DISCUSSION THREAD

**Discussion by jddmartin** [xournalpp/xournalpp#2024] [General]
Thank you xournalpp developers for a great program!

I was wondering what type of things *could* be done using the plug-in functionality --- the user manual doesn't have any mention of plug-ins.

I can see that there is an example in the repo --- but would writing a plug-in be possible without intimate knowledge of xournalpp internals?

Any general advice for someone who wanted to write plug-ins?

(The "feature request" here is more documentation on plug-ins.)

**Comment by LittleHuba** [comment]
@andreasb242 that's your field 😉

**Comment by rolandlo** [comment]
As an example: I have a plugin that takes a screenshot of a user selected region of the screen and pastes it into the document. It saves me some clicks. This plugin uses gnome-screenshot (which is available on Gnome-based Linux distros). It wasn't my idea, though, see [this post from user gyps](https://gitter.im/xournalpp/xournalpp?at=5d99e86937073b36a07e5e13). Just the idea to combine it with the pasting action was mine appearantly.

You could write plugins that use other system programs as well and combine them with actions that you can invoke from the menu.

**Comment by rolandlo** [comment]
You will now find some detailed explanations about how to create and use plugins on the work in progress website [section on plugins](https://xournalpp.github.io/guide/plugins/). 

Moreover in version `1.1.0+dev` we now have a plugin for cycling through a list of colors #2251 using a keyboard shortcut. Since you can also assign keyboard shortcuts to wacom buttons (using `xsetwacom` on Linux), you can make these actions activated by wacom buttons, for example. So plugins are quite useful for creating shortcuts for actions that you use often.

We also have a plugin for handwritten text recognition #2176 in preparation. It uses some Lua modules to create a more sophisticated UI and to make https-requests to a server, which does the actual work. That is also something you can do with plugins.

**Comment by rolandlo** [comment]
PR #2406 makes page and layer operations available to the plugin API. This is very useful if you want to apply some Xournal++ features to a (filtered) range of layers or pages, like changing the background of all pages that have non-standard page format or hiding all layers which have annotations.

The PR includes a plugin which allows autoscrolling. This may be useful if you use Xournal++ for displaying your music scores while playing your instrument.
It also includes a plugin which clones all handwritten layers to the next pages. This is useful for presentations with LaTeX Beamer and the like.
After the PR is merged, the work-in-progress website will have to be updated.

**Comment by rolandlo** [comment]
PR #2406 extending the Lua plugin API with layer and page operations is merged now. The API test plugin containing autoscrolling is still available for download there (it has not been merged). The [website](https://xournalpp.github.io/guide/plugins/) is updated to reflect this extension.
